kimchi/src/circuits/polynomials/keccak/witness.rs

```diff
@@ -31,15 +45,397 @@ fn _layout_round<F: PrimeField>() -> _Layout<F, KECCAK_COLS> {
+    expand_rot_e: Vec<u64>,
+    state_b: Vec<u64>,
+}
+
+impl PiRho {
+    fn create(state_e: &[u64]) -> Self {
+        let shifts_e = shift(state_e);
+        let dense_e = collapse(&reset(&shifts_e));
+        let rotation_e = Rotation::many(
+            &dense_e,
+            &OFF.iter()
+                .flatten()
+                .map(|x| *x as u32)
+                .collect::<Vec<u32>>(),
+        );
+
+        let mut state_b = vec![vec![vec![0; QUARTERS]; DIM]; DIM];
+        let aux = grid!(100, rotation_e.expand_rot);
+        for y in 0..DIM {
+            for x in 0..DIM {
+                for q in 0..QUARTERS {
+                    state_b[(2 * x + 3 * y) % DIM][y][q] = aux(y, x, q);
+                }
+            }
+        }
+        let state_b = state_b.iter().flatten().flatten().copied().collect();
+
+        Self {
+            shifts_e,
+            dense_e,
+            quotient_e: rotation_e.quotient,
+            remainder_e: rotation_e.remainder,
+            bound_e: rotation_e.bound,
+            dense_rot_e: rotation_e.dense_rot,
+            expand_rot_e: rotation_e.expand_rot,
+            state_b,
+        }
+    }
+}
+
+struct Chi {
+    shifts_b: Vec<u64>,
+    shifts_sum: Vec<u64>,
+    state_f: Vec<u64>,
+}
+
+impl Chi {
+    fn create(state_b: &[u64]) -> Self {
+        let shifts_b = shift(state_b);
+        let shiftsb = grid!(400, shifts_b);
+        let mut sum = vec![];
+        for y in 0..DIM {
+            for x in 0..DIM {
+                for q in 0..QUARTERS {
+                    let not = 0x1111111111111111u64 - shiftsb(0, y, (x + 1) % DIM, q);
+                    sum.push(not + shiftsb(0, y, (x + 2) % DIM, q));
+                }
+            }
+        }
+        let shifts_sum = shift(&sum);
+        let shiftsum = grid!(400, shifts_sum);
+        let mut state_f = vec![];
+        for y in 0..DIM {
+            for x in 0..DIM {
+                for q in 0..QUARTERS {
+                    let and = shiftsum(1, y, x, q);
+                    state_f.push(shiftsb(0, y, x, q) + and);
+                }
+            }
+        }
+
+        Self {
+            shifts_b,
+            shifts_sum,
+            state_f,
+        }
+    }
+}
+
+struct Iota {
+    state_g: Vec<u64>,
+}
+
+impl Iota {
+    fn create(state_f: Vec<u64>, round: usize) -> Self {
+        let rc = decompose(RC[round])
+            .iter()
+            .map(|x| expand(*x))
+            .collect::<Vec<u64>>();
+        let mut state_g = state_f.clone();
+        for (i, c) in rc.iter().enumerate() {
+            state_g[i] = state_f[i] + *c;
+        }
+        Self { state_g }
+    }
+}
+
+/// Creates a witness for the Keccak hash function
+/// Input:
+/// - message: the message to be hashed
+/// Note:
+/// Requires at least one more row after the keccak gadget so that
+/// constraints can access the next row in the squeeze
+pub fn extend_keccak_witness<F: PrimeField>(witness: &mut [Vec<F>; KECCAK_COLS], message: BigUint) {
```

Comment:
Nit: there are constraints on the field size. But I don't know if we want to ensure that in the code.